Regression with Gradle's configuration cache in 0.9.15

**Describe the bug**
In some circumstances, a Gradle build fails when run using `--configuration-cache` due to a `Configuration` instance that's captured by a lambda in `NativeImpagePlugin`. The same build works with 0.9.14.

**To Reproduce**
Sorry, I've yet to manage to boil things down into a minimal reproducer. Hopefully the configuration cache problem report (see below) will be sufficient to diagnose the problem.

**Expected behavior**
The Gradle plugin should work with `--configuration-cache`.

**Logs**

```
⌄ cannot serialize object of type org.gradle.api.internal.artifacts.configurations.DefaultConfiguration, a subtype of org.gradle.api.artifacts.Configuration, as these are not supported with the configuration cache. ?
  ⌄ fieldarg$1 of org.graalvm.buildtools.gradle.NativeImagePlugin$$Lambda$1852/0x0000000801a9f000
    ⌄ bean of type org.graalvm.buildtools.gradle.NativeImagePlugin$$Lambda$1852/0x0000000801a9f000
      ⌄ fieldtransformer of org.gradle.api.internal.provider.TransformBackedProvider
        ⌄ bean of type org.gradle.api.internal.provider.TransformBackedProvider
          ⌄ fieldprovider of org.gradle.configurationcache.serialization.codecs.ProviderBackedFileCollectionSpec
            ⌄ bean of type org.gradle.configurationcache.serialization.codecs.ProviderBackedFileCollectionSpec
              ⌄ field__configurationFileDirectories__ of org.graalvm.buildtools.gradle.internal.BaseNativeImageOptions
                ⌄ bean of type org.graalvm.buildtools.gradle.internal.BaseNativeImageOptions
                  ⌄ field_gr_owner_ of org.graalvm.buildtools.gradle.dsl.NativeResourcesOptions
                    ⌄ bean of type org.graalvm.buildtools.gradle.dsl.NativeResourcesOptions
                      ⌄ field__options__ of org.graalvm.buildtools.gradle.tasks.GenerateResourcesConfigFile
                        task:generateResourcesConfigFile of type org.graalvm.buildtools.gradle.tasks.GenerateResourcesConfigFile
```

**System Info (please complete the following information):**
 - OS: macOS
 - GraalVM Version: 22.2
 - Java Version: 17
 - Plugin version: 0.9.15

That was quick. Thank you, @melix. That's better as the configuration cache no longer reports a problem. Unfortunately, the build still fails:

```
./gradlew --configuration-cache generateResourcesConfigFile
Configuration cache is an incubating feature.
Calculating task graph as no configuration cache is available for tasks: generateResourcesConfigFile

0 problems were found storing the configuration cache.

See the complete report at file:///Users/awilkinson/dev/temp/nbt-configuration-cache-problem/build/reports/configuration-cache/bub752nr41ig1gyl5xvezwkms/cku5jhqpzmyxpzegunrchhv9/configuration-cache-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Configuration cache state could not be cached: field '__options__' from type 'org.graalvm.buildtools.gradle.tasks.GenerateResourcesConfigFile': error writing value of type 'org.gradle.api.internal.provider.DefaultProperty'
> Configuration cache state could not be cached: field '_gr_owner_' from type 'org.graalvm.buildtools.gradle.dsl.NativeResourcesOptions': error writing value of type 'org.graalvm.buildtools.gradle.internal.BaseNativeImageOptions'
   > Configuration cache state could not be cached: field '__configurationFileDirectories__' from type 'org.graalvm.buildtools.gradle.internal.BaseNativeImageOptions': error writing value of type 'org.gradle.api.internal.file.collections.DefaultConfigurableFileCollection'
      > Configuration cache state could not be cached: field 'provider' from type 'org.gradle.configurationcache.serialization.codecs.ProviderBackedFileCollectionSpec': error writing value of type 'org.gradle.api.internal.provider.FlatMapProvider'
         > Configuration cache state could not be cached: field 'transformer' from type 'org.gradle.api.internal.provider.TransformBackedProvider': error writing value of type 'org.graalvm.buildtools.gradle.NativeImagePlugin$$Lambda$1256/0x000000080155dc28'
            > Configuration cache state could not be cached: field 'capturedArgs' from type 'java.lang.invoke.SerializedLambda': error writing value of type '[Ljava.lang.Object;'
               > Unable to make field private final java.util.function.BiConsumer java.util.stream.Collectors$CollectorImpl.accumulator accessible: module java.base does not "opens java.util.stream" to unnamed module @bbf12e9

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s
Configuration cache entry discarded.
```

I don't see the failure with 0.9.14. It occurs with 0.9.15 and 0.9.16-SNAPSHOT.

[nbt-configuration-cache-problem.zip](https://github.com/graalvm/native-build-tools/files/9801134/nbt-configuration-cache-problem.zip) contains a reproducer that should fail as above.

--- a/.github/workflows/test-native-gradle-plugin.yml
+++ b/.github/workflows/test-native-gradle-plugin.yml
@@ -53,7 +53,8 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        gradle-version: ["current", "7.1"]
+        gradle-version: ["current", "7.4"]
+        gradle-config-cache-version: ["current", "7.5.1"]
         # Following versions are disabled temporarily in order to speed up PR testing
         # "7.3.3", "7.2", "7.1", "6.8.3"
         graalvm-version: [ latest ] # dev
@@ -69,6 +70,8 @@ jobs:
           github-token: ${{ secrets.GITHUB_TOKEN }}
       - name: "❓ Check and test the plugin"
         run: ./gradlew :native-gradle-plugin:functionalTest -DgradleVersion=${{ matrix.gradle-version }}
+      - name: "❓ Check and test the plugin with configuration cache"
+        run: ./gradlew :native-gradle-plugin:configCacheFunctionalTest -DgradleVersion=${{ matrix.gradle-config-cache-version }}
       - name: "📜 Upload functional tests results"
         if: always()
         uses: actions/upload-artifact@v2

--- a/build-logic/gradle-functional-testing/src/main/groovy/org.graalvm.build.functional-testing.gradle
+++ b/build-logic/gradle-functional-testing/src/main/groovy/org.graalvm.build.functional-testing.gradle
@@ -82,12 +82,12 @@ def graalVm = javaToolchains.launcherFor {
 def fullFunctionalTest = tasks.register("fullFunctionalTest")
 
 ['functionalTest', 'configCacheFunctionalTest'].each { baseName ->
-    ["current", "7.3.3", "7.2", "7.1"].each { gradleVersion ->
+    ["current", "7.4", "7.6.2"].each { gradleVersion ->
         String taskName = gradleVersion == 'current' ? baseName : "gradle${gradleVersion}${baseName.capitalize()}"
         // Add a task to run the functional tests
         def testTask = tasks.register(taskName, Test) {
             String effectiveVersion = gradleVersion
-            def versionProvider = providers.systemProperty('gradleVersion').forUseAtConfigurationTime()
+            def versionProvider = providers.systemProperty('gradleVersion')
             if (effectiveVersion == 'current' && versionProvider.isPresent()) {
                 effectiveVersion = versionProvider.get()
             }
@@ -99,7 +99,7 @@ def fullFunctionalTest = tasks.register("fullFunctionalTest")
             systemProperty('common.repo.url', configurations.functionalTestCommonRepository.incoming.files.files[0])
             systemProperty('gradle.test.version', effectiveVersion)
             systemProperty('versions.junit', libs.versions.junitJupiter.get())
-            environment('GRAALVM_HOME', graalVm.forUseAtConfigurationTime().get().metadata.installationPath.asFile.absolutePath)
+            environment('GRAALVM_HOME', graalVm.get().metadata.installationPath.asFile.absolutePath)
             testClassesDirs = sourceSets.functionalTest.output.classesDirs
             classpath = sourceSets.functionalTest.runtimeClasspath
             javaLauncher.set(graalVm)

--- a/common/graalvm-reachability-metadata/gradle/native-image-testing.gradle
+++ b/common/graalvm-reachability-metadata/gradle/native-image-testing.gradle
@@ -136,7 +136,7 @@ tasks.register("nativeTestCompile", Exec) {
     argsProvider.classpath.from(test.classpath)
     argsProvider.testIdsDir.set(testIdsDir)
     argsProvider.agentOutputDir.set(agentOutput)
-    argsProvider.discovery.set(providers.systemProperty("testDiscovery").forUseAtConfigurationTime().map(v -> Boolean.valueOf(v)).orElse(false))
+    argsProvider.discovery.set(providers.systemProperty("testDiscovery").map(v -> Boolean.valueOf(v)).orElse(false))
     argumentProviders.add(argsProvider)
 }
 

--- a/common/junit-platform-native/gradle/native-image-testing.gradle
+++ b/common/junit-platform-native/gradle/native-image-testing.gradle
@@ -136,7 +136,7 @@ tasks.register("nativeTestCompile", Exec) {
     argsProvider.classpath.from(test.classpath)
     argsProvider.testIdsDir.set(testIdsDir)
     argsProvider.agentOutputDir.set(agentOutput)
-    argsProvider.discovery.set(providers.systemProperty("testDiscovery").forUseAtConfigurationTime().map(v -> Boolean.valueOf(v)).orElse(false))
+    argsProvider.discovery.set(providers.systemProperty("testDiscovery").map(v -> Boolean.valueOf(v)).orElse(false))
     argumentProviders.add(argsProvider)
 }
 

--- a/docs/build.gradle.kts
+++ b/docs/build.gradle.kts
@@ -44,7 +44,6 @@ plugins {
 }
 
 version = providers.gradleProperty("publishVersion")
-        .forUseAtConfigurationTime()
         .orElse(libs.versions.nativeBuildTools)
         .get()
 

--- a/docs/src/docs/asciidoc/index.adoc
+++ b/docs/src/docs/asciidoc/index.adoc
@@ -21,6 +21,12 @@ If you are using alternative build systems, see <<alternative-build-systems.adoc
 
 === Release 0.9.21
 
+==== Gradle plugin
+
+- Bump minimal version of Gradle to 7.4
+- Fix compatibility with Gradle's https://docs.gradle.org/8.0.2/userguide/configuration_cache.html#header[configuration cache] (requires Gradle 7.5+)
+- Remove use of deprecated Gradle APIs
+
 ==== Maven plugin
 
 - Add a new `native:write-args-file` goal that can be used to write the arguments passed to `native-image` to a file

--- a/gradle/wrapper/gradle-wrapper.properties
+++ b/gradle/wrapper/gradle-wrapper.properties
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-7.4-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.0.2-bin.zip
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists

--- a/native-gradle-plugin/src/functionalTest/groovy/org/graalvm/buildtools/gradle/DeprecatedExtensionFunctionalTest.groovy
+++ b/native-gradle-plugin/src/functionalTest/groovy/org/graalvm/buildtools/gradle/DeprecatedExtensionFunctionalTest.groovy
@@ -42,8 +42,10 @@
 package org.graalvm.buildtools.gradle
 
 import org.graalvm.buildtools.gradle.fixtures.AbstractFunctionalTest
+import spock.lang.IgnoreIf
 
 class DeprecatedExtensionFunctionalTest extends AbstractFunctionalTest {
+    @IgnoreIf({ Boolean.getBoolean("config.cache")} )
     def "using a deprecated extension issues a warning"() {
         given:
         buildFile << """

--- a/native-gradle-plugin/src/functionalTest/groovy/org/graalvm/buildtools/gradle/JavaApplicationFunctionalTest.groovy
+++ b/native-gradle-plugin/src/functionalTest/groovy/org/graalvm/buildtools/gradle/JavaApplicationFunctionalTest.groovy
@@ -43,6 +43,7 @@ package org.graalvm.buildtools.gradle
 
 import org.graalvm.buildtools.gradle.fixtures.AbstractFunctionalTest
 import org.gradle.util.GradleVersion
+import spock.lang.IgnoreIf
 import spock.lang.Issue
 import spock.lang.Requires
 
@@ -114,6 +115,11 @@ class JavaApplicationFunctionalTest extends AbstractFunctionalTest {
         def graalvmHome = System.getenv("GRAALVM_HOME")
         graalvmHome != null
     })
+    @IgnoreIf({
+        // Incompatible with configuration cache because the toolchain error causes serialization
+        // problems
+        Boolean.getBoolean("config.cache")
+    })
     def "can override toolchain selection"() {
         def nativeApp = file("build/native/nativeCompile/java-application")
         boolean dummyToolchain = GradleVersion.version(gradleVersion).compareTo(GradleVersion.version("7.0")) >= 0
@@ -124,6 +130,7 @@ class JavaApplicationFunctionalTest extends AbstractFunctionalTest {
         if (dummyToolchain) {
             buildFile << """graalvmNative.binaries.configureEach {
                 javaLauncher.set(javaToolchains.launcherFor {
+                    languageVersion.set(JavaLanguageVersion.of(JavaVersion.current().getMajorVersion()))
                     vendor.set(JvmVendorSpec.matching("non existing vendor"))
                 })
             }"""

--- a/native-gradle-plugin/src/functionalTest/groovy/org/graalvm/buildtools/gradle/JavaApplicationFunctionalTest.groovy
+++ b/native-gradle-plugin/src/functionalTest/groovy/org/graalvm/buildtools/gradle/JavaApplicationFunctionalTest.groovy
@@ -42,8 +42,7 @@
 package org.graalvm.buildtools.gradle
 
 import org.graalvm.buildtools.gradle.fixtures.AbstractFunctionalTest
-import org.gradle.util.GradleVersion
-import spock.lang.IgnoreIf
+import spock.lang.Ignore
 import spock.lang.Issue
 import spock.lang.Requires
 
@@ -115,28 +114,20 @@ class JavaApplicationFunctionalTest extends AbstractFunctionalTest {
         def graalvmHome = System.getenv("GRAALVM_HOME")
         graalvmHome != null
     })
-    @IgnoreIf({
-        // Incompatible with configuration cache because the toolchain error causes serialization
-        // problems
-        Boolean.getBoolean("config.cache")
-    })
+    @Ignore("Need to find another way to test this since toolchains will now always be evaluated early")
     def "can override toolchain selection"() {
         def nativeApp = file("build/native/nativeCompile/java-application")
-        boolean dummyToolchain = GradleVersion.version(gradleVersion).compareTo(GradleVersion.version("7.0")) >= 0
 
         given:
         withSample("java-application")
 
-        if (dummyToolchain) {
             buildFile << """graalvmNative.binaries.configureEach {
                 javaLauncher.set(javaToolchains.launcherFor {
                     languageVersion.set(JavaLanguageVersion.of(JavaVersion.current().getMajorVersion()))
                     vendor.set(JvmVendorSpec.matching("non existing vendor"))
                 })
-            }"""
-        }
-
-        buildFile << """
+            }
+            
             tasks.withType(org.graalvm.buildtools.gradle.tasks.BuildNativeImageTask).configureEach {
                 disableToolchainDetection = true
             }

--- a/native-gradle-plugin/src/functionalTest/groovy/org/graalvm/buildtools/gradle/JavaApplicationWithTestsFunctionalTest.groovy
+++ b/native-gradle-plugin/src/functionalTest/groovy/org/graalvm/buildtools/gradle/JavaApplicationWithTestsFunctionalTest.groovy
@@ -284,7 +284,12 @@ class JavaApplicationWithTestsFunctionalTest extends AbstractFunctionalTest {
 
         then:
         tasks {
-            upToDate(':test')
+            if (hasConfigurationCache) {
+                // because we run with --rerun-tasks when checking with config cache
+                succeeded(':test')
+            } else {
+                upToDate(':test')
+            }
         }
     }
 }

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/NativeImagePlugin.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/NativeImagePlugin.java
@@ -139,7 +139,6 @@ import java.util.stream.Collector;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
-import static org.graalvm.buildtools.gradle.internal.ConfigurationCacheSupport.serializableBiFunctionOf;
 import static org.graalvm.buildtools.gradle.internal.ConfigurationCacheSupport.serializableFunctionOf;
 import static org.graalvm.buildtools.gradle.internal.ConfigurationCacheSupport.serializablePredicateOf;
 import static org.graalvm.buildtools.gradle.internal.ConfigurationCacheSupport.serializableSupplierOf;
@@ -330,7 +329,7 @@ public class NativeImagePlugin implements Plugin<Project> {
             tasks.register(runTaskName, NativeRunTask.class, task -> {
                 task.setGroup(LifecycleBasePlugin.BUILD_GROUP);
                 task.setDescription("Executes the " + options.getName() + " native binary");
-                task.getImage().convention(imageBuilder.map(t -> t.getOutputFile().get()));
+                task.getImage().convention(imageBuilder.flatMap(BuildNativeImageTask::getOutputFile));
                 task.getRuntimeArgs().convention(options.getRuntimeArgs());
             });
             configureClasspathJarFor(tasks, options, imageBuilder);
@@ -374,13 +373,15 @@ public class NativeImagePlugin implements Plugin<Project> {
                                                            Collector<T, A, R> collector) {
         GraalVMReachabilityMetadataRepositoryExtension extension = reachabilityExtensionOn(graalExtension);
         Provider<GraalVMReachabilityMetadataService> metadataServiceProvider = graalVMReachabilityMetadataService(project, extension);
-        Configuration classpath = project.getConfigurations().getByName(sourceSet.getRuntimeClasspathConfigurationName());
+        ResolutionResult resolutionResult = project.getConfigurations()
+                .getByName(sourceSet.getRuntimeClasspathConfigurationName())
+                .getIncoming()
+                .getResolutionResult();
         return extension.getEnabled().flatMap(serializableTransformerOf(enabled -> {
             if (enabled && extension.getUri().isPresent()) {
                 Set<String> excludedModules = extension.getExcludedModules().getOrElse(Collections.emptySet());
                 Map<String, String> forcedVersions = extension.getModuleToConfigVersion().getOrElse(Collections.emptyMap());
                 return metadataServiceProvider.map(serializableTransformerOf(service -> {
-                    ResolutionResult resolutionResult = classpath.getIncoming().getResolutionResult();
                     Set<ResolvedComponentResult> components = resolutionResult.getAllComponents();
                     Stream<T> mapped = components.stream().flatMap(serializableFunctionOf(component -> {
                         ModuleVersionIdentifier moduleVersion = component.getModuleVersion();
@@ -441,9 +442,14 @@ public class NativeImagePlugin implements Plugin<Project> {
     }
 
     private void configureJvmReachabilityExcludeConfigArgs(Project project, GraalVMExtension graalExtension, NativeImageOptions options, SourceSet sourceSet) {
-        options.getExcludeConfig().putAll(graalVMReachabilityQuery(project, graalExtension, sourceSet,
-                DirectoryConfiguration::isOverride, this::getExclusionConfig,
-                () -> Collectors.<Map.Entry<String, List<String>>, String, List<String>>toMap(Map.Entry::getKey, Map.Entry::getValue)));
+        options.getExcludeConfig().putAll(
+                graalVMReachabilityQuery(project,
+                        graalExtension,
+                        sourceSet,
+                        DirectoryConfiguration::isOverride,
+                        this::getExclusionConfig,
+                        Collectors.<Map.Entry<String, List<String>>, String, List<String>>toMap(Map.Entry::getKey, Map.Entry::getValue))
+        );
         GraalVMReachabilityMetadataRepositoryExtension repositoryExtension = reachabilityExtensionOn(graalExtension);
         graalVMReachabilityMetadataService(project, repositoryExtension);
     }
@@ -627,7 +633,6 @@ public class NativeImagePlugin implements Plugin<Project> {
     private static Provider<String> agentProperty(Project project, AgentOptions options) {
         return project.getProviders()
                 .gradleProperty(AGENT_PROPERTY)
-                .forUseAtConfigurationTime()
                 .map(serializableTransformerOf(v -> {
                     if (!v.isEmpty()) {
                         return v;

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/NativeImagePlugin.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/NativeImagePlugin.java
@@ -78,6 +78,7 @@ import org.gradle.api.artifacts.ConfigurationContainer;
 import org.gradle.api.artifacts.Dependency;
 import org.gradle.api.artifacts.ModuleVersionIdentifier;
 import org.gradle.api.artifacts.component.ModuleComponentIdentifier;
+import org.gradle.api.artifacts.result.ResolutionResult;
 import org.gradle.api.artifacts.result.ResolvedComponentResult;
 import org.gradle.api.attributes.Attribute;
 import org.gradle.api.attributes.AttributeContainer;
@@ -138,6 +139,8 @@ import java.util.stream.Collector;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
+import static org.graalvm.buildtools.gradle.internal.ConfigurationCacheSupport.serializableBiFunctionOf;
+import static org.graalvm.buildtools.gradle.internal.ConfigurationCacheSupport.serializableFunctionOf;
 import static org.graalvm.buildtools.gradle.internal.ConfigurationCacheSupport.serializablePredicateOf;
 import static org.graalvm.buildtools.gradle.internal.ConfigurationCacheSupport.serializableSupplierOf;
 import static org.graalvm.buildtools.gradle.internal.ConfigurationCacheSupport.serializableTransformerOf;
@@ -288,12 +291,12 @@ public class NativeImagePlugin implements Plugin<Project> {
         GraalVMReachabilityMetadataRepositoryExtension metadataRepositoryExtension = reachabilityExtensionOn(graalExtension);
         TaskCollection<CollectReachabilityMetadata> reachabilityMetadataCopyTasks = project.getTasks()
                 .withType(CollectReachabilityMetadata.class);
-         reachabilityMetadataCopyTasks.configureEach(task -> {
-             Provider<GraalVMReachabilityMetadataService> reachabilityMetadataService = graalVMReachabilityMetadataService(
-                     project, metadataRepositoryExtension);
+        reachabilityMetadataCopyTasks.configureEach(task -> {
+            Provider<GraalVMReachabilityMetadataService> reachabilityMetadataService = graalVMReachabilityMetadataService(
+                    project, metadataRepositoryExtension);
             task.getMetadataService().set(reachabilityMetadataService);
             task.usesService(reachabilityMetadataService);
-            task.getUri().convention(task.getVersion().map(this::getReachabilityMetadataRepositoryUrlForVersion)
+            task.getUri().convention(task.getVersion().map(serializableTransformerOf(this::getReachabilityMetadataRepositoryUrlForVersion))
                     .orElse(metadataRepositoryExtension.getUri()));
             task.getExcludedModules().convention(metadataRepositoryExtension.getExcludedModules());
             task.getModuleToConfigVersion().convention(metadataRepositoryExtension.getModuleToConfigVersion());
@@ -338,47 +341,57 @@ public class NativeImagePlugin implements Plugin<Project> {
                     tasks,
                     transitiveProjectArtifacts(project, sourceSet.getRuntimeClasspathConfigurationName()),
                     deriveTaskName(binaryName, "generate", "ResourcesConfigFile"));
-            options.getConfigurationFileDirectories().from(generateResourcesConfig.map(t ->
+            options.getConfigurationFileDirectories().from(generateResourcesConfig.map(serializableTransformerOf(t ->
                     t.getOutputFile().map(f -> f.getAsFile().getParentFile())
-            ));
+            )));
             configureJvmReachabilityConfigurationDirectories(project, graalExtension, options, sourceSet);
             configureJvmReachabilityExcludeConfigArgs(project, graalExtension, options, sourceSet);
         });
     }
 
-    private void configureJvmReachabilityConfigurationDirectories(Project project, GraalVMExtension graalExtension,
-            NativeImageOptions options, SourceSet sourceSet) {
-        options.getConfigurationFileDirectories().from(graalVMReachabilityQuery(project, graalExtension, sourceSet,
-                        configuration -> true, this::getConfigurationDirectory,
-                        Collectors.toList()));
+    private void configureJvmReachabilityConfigurationDirectories(Project project,
+                                                                  GraalVMExtension graalExtension,
+                                                                  NativeImageOptions options,
+                                                                  SourceSet sourceSet) {
+        options.getConfigurationFileDirectories().from(
+                graalVMReachabilityQuery(project, graalExtension, sourceSet,
+                        serializablePredicateOf(configuration -> true),
+                        serializableBiFunctionOf(this::getConfigurationDirectory),
+                        Collectors::<File>toList)
+        );
     }
 
     private File getConfigurationDirectory(ModuleVersionIdentifier moduleVersion,
-            DirectoryConfiguration configuration) {
+                                           DirectoryConfiguration configuration) {
         return configuration.getDirectory().toAbsolutePath().toFile();
     }
 
     private <T, A, R> Provider<R> graalVMReachabilityQuery(Project project, GraalVMExtension graalExtension,
-            SourceSet sourceSet, Predicate<DirectoryConfiguration> filter,
-            BiFunction<ModuleVersionIdentifier, DirectoryConfiguration, T> mapper, Collector<T, A, R> collector) {
+                                                           SourceSet sourceSet, Predicate<DirectoryConfiguration> filter,
+                                                           BiFunction<ModuleVersionIdentifier, DirectoryConfiguration, T> mapper,
+                                                           Supplier<Collector<T, A, R>> collector) {
         GraalVMReachabilityMetadataRepositoryExtension extension = reachabilityExtensionOn(graalExtension);
-        return extension.getEnabled().flatMap(enabled -> {
+        Provider<GraalVMReachabilityMetadataService> metadataServiceProvider = graalVMReachabilityMetadataService(project, extension);
+        Configuration classpath = project.getConfigurations().getByName(sourceSet.getRuntimeClasspathConfigurationName());
+        return extension.getEnabled().flatMap(serializableTransformerOf(enabled -> {
             if (enabled && extension.getUri().isPresent()) {
-                Configuration classpath = project.getConfigurations().getByName(sourceSet.getRuntimeClasspathConfigurationName());
                 Set<String> excludedModules = extension.getExcludedModules().getOrElse(Collections.emptySet());
                 Map<String, String> forcedVersions = extension.getModuleToConfigVersion().getOrElse(Collections.emptyMap());
-                return graalVMReachabilityMetadataService(project, extension).map(service -> {
-                    Set<ResolvedComponentResult> components = classpath.getIncoming().getResolutionResult().getAllComponents();
-                    Stream<T> mapped = components.stream().flatMap(component -> {
+                return metadataServiceProvider.map(serializableTransformerOf(service -> {
+                    ResolutionResult resolutionResult = classpath.getIncoming().getResolutionResult();
+                    Set<ResolvedComponentResult> components = resolutionResult.getAllComponents();
+                    Stream<T> mapped = components.stream().flatMap(serializableFunctionOf(component -> {
                         ModuleVersionIdentifier moduleVersion = component.getModuleVersion();
                         Set<DirectoryConfiguration> configurations = service.findConfigurationsFor(excludedModules, forcedVersions, moduleVersion);
-                        return configurations.stream().filter(filter).map(configuration -> mapper.apply(moduleVersion, configuration));
-                    });
-                    return mapped.collect(collector);
-                });
+                        return configurations.stream()
+                                .filter(filter)
+                                .map(serializableFunctionOf(configuration -> mapper.apply(moduleVersion, configuration)));
+                    }));
+                    return mapped.collect(collector.get());
+                }));
             }
-            return project.getProviders().provider(() -> Stream.<T>empty().collect(collector));
-        });
+            return project.getProviders().provider(() -> Stream.<T>empty().collect(collector.get()));
+        }));
     }
 
     private Provider<GraalVMReachabilityMetadataService> graalVMReachabilityMetadataService(Project project,
@@ -388,7 +401,7 @@ public class NativeImagePlugin implements Plugin<Project> {
                 .registerIfAbsent("nativeConfigurationService", GraalVMReachabilityMetadataService.class, spec -> {
                     LogLevel logLevel = determineLogLevel();
                     spec.getParameters().getLogLevel().set(logLevel);
-                    spec.getParameters().getUri().set(repositoryExtension.getUri().map(configuredUri -> computeMetadataRepositoryUri(project, repositoryExtension, configuredUri, GraalVMLogger.of(project.getLogger()))));
+                    spec.getParameters().getUri().set(repositoryExtension.getUri().map(serializableTransformerOf(configuredUri -> computeMetadataRepositoryUri(project, repositoryExtension, configuredUri, GraalVMLogger.of(project.getLogger())))));
                     spec.getParameters().getCacheDir().set(
                             new File(project.getGradle().getGradleUserHomeDir(), "native-build-tools/repositories"));
                 });
@@ -418,7 +431,7 @@ public class NativeImagePlugin implements Plugin<Project> {
                 return files.iterator().next().toURI();
             } else {
                 logger.warn("Unable to find the GraalVM reachability metadata repository in Maven repository. " +
-                        "Falling back to the default repository at " + defaultUri);
+                            "Falling back to the default repository at " + defaultUri);
             }
         }
         return configuredUri;
@@ -427,7 +440,7 @@ public class NativeImagePlugin implements Plugin<Project> {
     private void configureJvmReachabilityExcludeConfigArgs(Project project, GraalVMExtension graalExtension, NativeImageOptions options, SourceSet sourceSet) {
         options.getExcludeConfig().putAll(graalVMReachabilityQuery(project, graalExtension, sourceSet,
                 DirectoryConfiguration::isOverride, this::getExclusionConfig,
-                Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue)));
+                () -> Collectors.<Map.Entry<String, List<String>>, String, List<String>>toMap(Map.Entry::getKey, Map.Entry::getValue)));
         GraalVMReachabilityMetadataRepositoryExtension repositoryExtension = reachabilityExtensionOn(graalExtension);
         project.getGradle()
                 .getSharedServices()
@@ -524,7 +537,7 @@ public class NativeImagePlugin implements Plugin<Project> {
         GraalVMReachabilityMetadataRepositoryExtension configurationRepository = graalvmNative.getExtensions().create("metadataRepository", GraalVMReachabilityMetadataRepositoryExtension.class);
         configurationRepository.getEnabled().convention(false);
         configurationRepository.getVersion().convention(VersionInfo.METADATA_REPO_VERSION);
-        configurationRepository.getUri().convention(configurationRepository.getVersion().map(this::getReachabilityMetadataRepositoryUrlForVersion));
+        configurationRepository.getUri().convention(configurationRepository.getVersion().map(serializableTransformerOf(this::getReachabilityMetadataRepositoryUrlForVersion)));
         configurationRepository.getExcludedModules().convention(Collections.emptySet());
         configurationRepository.getModuleToConfigVersion().convention(Collections.emptyMap());
     }
@@ -721,7 +734,7 @@ public class NativeImagePlugin implements Plugin<Project> {
     }
 
     private static void setupExtensionConfigExcludes(NativeImageOptions options, NativeConfigurations configs) {
-        options.getExcludeConfigArgs().set(options.getExcludeConfig().map(excludedConfig -> {
+        options.getExcludeConfigArgs().set(options.getExcludeConfig().map(serializableTransformerOf(excludedConfig -> {
             List<String> args = new ArrayList<>();
             excludedConfig.forEach((entry, listOfResourcePatterns) -> {
                 if (entry instanceof File) {
@@ -751,7 +764,7 @@ public class NativeImagePlugin implements Plugin<Project> {
                 }
             });
             return args;
-        }));
+        })));
     }
 
     private static List<String> agentSessionDirectories(Directory outputDirectory) {

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/NativeImageService.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/NativeImageService.java
@@ -66,8 +66,7 @@ public abstract class NativeImageService implements BuildService<BuildServicePar
 
     private static Provider<Integer> maxParallelUsagesOf(ProviderFactory providers) {
         return providers.gradleProperty(MAX_PARALLEL_SYSTEM_PROPERTY)
-                .forUseAtConfigurationTime()
-                .orElse(providers.environmentVariable(MAX_PARALLEL_ENV_VAR).forUseAtConfigurationTime())
+                .orElse(providers.environmentVariable(MAX_PARALLEL_ENV_VAR))
                 .map(Integer::parseInt)
                 .orElse(1 + Runtime.getRuntime().availableProcessors() / 16);
     }

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/ConfigurationCacheSupport.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/ConfigurationCacheSupport.java
@@ -44,8 +44,11 @@ package org.graalvm.buildtools.gradle.internal;
 import org.gradle.api.Transformer;
 
 import java.io.Serializable;
+import java.util.function.BiFunction;
+import java.util.function.Function;
 import java.util.function.Predicate;
 import java.util.function.Supplier;
+import java.util.stream.Collector;
 
 /**
  * Helper class to deal with Gradle configuration cache.
@@ -82,6 +85,33 @@ public class ConfigurationCacheSupport {
         return transformer;
     }
 
+    /**
+     * Generates a serializable function lambda.
+     * @param function the bifunction
+     * @param <F> the type of the parameter
+     * @param <T> the type of the result
+     * @return a serializable function
+     */
+    public static <F, T> Function<F, T> serializableFunctionOf(SerializableFunction<F, T> function) {
+        return function;
+    }
+
+    /**
+     * Generates a serializable bifunction lambda.
+     * @param bifunction the bifunction
+     * @param <T> the type of the first parameter
+     * @param <U> the type of the second parameter
+     * @param <R> the type of the result
+     * @return a serializable bifunction
+     */
+    public static <T, U, R> BiFunction<T, U, R> serializableBiFunctionOf(SerializableBiFunction<T, U, R> bifunction) {
+        return bifunction;
+    }
+
+    public static <T, A, R> Collector<T, A, R> serializableCollectorOf(SerializableCollector<T, A, R> collector) {
+        return collector;
+    }
+
     public interface SerializableSupplier<T> extends Supplier<T>, Serializable {
 
     }
@@ -91,6 +121,18 @@ public class ConfigurationCacheSupport {
     }
 
     public interface SerializableTransformer<OUT, IN> extends Transformer<OUT, IN>, Serializable {
+
+    }
+
+    public interface SerializableFunction<F, T> extends Function<F, T>, Serializable {
+
+    }
+
+    public interface SerializableBiFunction<T, U, R> extends BiFunction<T, U, R>, Serializable {
+
+    }
+
+    public interface SerializableCollector<T, A, R> extends Collector<T, A, R>, Serializable {
 
     }
 

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/DefaultGraalVmExtension.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/DefaultGraalVmExtension.java
@@ -97,9 +97,7 @@ public abstract class DefaultGraalVmExtension implements GraalVMExtension {
                         if (toolchainService != null) {
                             return toolchainService.launcherFor(spec -> {
                                 spec.getLanguageVersion().set(JavaLanguageVersion.of(JavaVersion.current().getMajorVersion()));
-                                if (GradleUtils.isAtLeastGradle7()) {
-                                    spec.getVendor().set(JvmVendorSpec.matching("GraalVM"));
-                                }
+                                spec.getVendor().set(JvmVendorSpec.matching("GraalVM"));
                             });
                         }
                     }

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/GradleUtils.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/GradleUtils.java
@@ -57,9 +57,7 @@ import org.gradle.util.GradleVersion;
  */
 @SuppressWarnings("unused")
 public class GradleUtils {
-    private static final GradleVersion GRADLE_68 = GradleVersion.version("6.8");
-    private static final GradleVersion GRADLE_7 = GradleVersion.version("7.0");
-    private static final GradleVersion GRADLE_71 = GradleVersion.version("7.1");
+    private static final GradleVersion MINIMAL_GRADLE_VERSION = GradleVersion.version("7.4");
 
     public static SourceSet findSourceSet(Project project, String sourceSetName) {
         SourceSetContainer sourceSetContainer = getJavaPluginConvention(project).getSourceSets();
@@ -70,17 +68,6 @@ public class GradleUtils {
         return project.getConvention().getPlugin(JavaPluginConvention.class);
     }
 
-    public static boolean isAtLeastGradle7() {
-        return GradleVersion.current().compareTo(GRADLE_7) >= 0;
-    }
-
-    public static boolean isAtLeastGradle6dot8() {
-        return GradleVersion.current().compareTo(GRADLE_68) >= 0;
-    }
-
-    public static boolean isAtLeastGradle7dot1() {
-        return GradleVersion.current().compareTo(GRADLE_71) >= 0;
-    }
 
     public static Configuration findConfiguration(Project project, String name) {
         return project.getConfigurations().getByName(name);

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/NativeImageExecutableLocator.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/NativeImageExecutableLocator.java
@@ -128,7 +128,6 @@ public class NativeImageExecutableLocator {
         public Provider<String> fromEnvVar(String envVar, ProviderFactory factory) {
             return factory.environmentVariable(envVar)
                     // required for older Gradle versions support
-                    .forUseAtConfigurationTime()
                     .map(ConfigurationCacheSupport.serializableTransformerOf(value -> {
                         this.envVar = envVar;
                         return value;
@@ -164,12 +163,8 @@ public class NativeImageExecutableLocator {
             if (toolchain != null) {
                 diags.add("GraalVM uses toolchain detection. Selected:");
                 diags.add("   - language version: " + toolchain.getLanguageVersion());
-                if (GradleUtils.isAtLeastGradle6dot8()) {
-                    diags.add("   - vendor: " + toolchain.getVendor());
-                    if (GradleUtils.isAtLeastGradle7dot1()) {
-                        diags.add("   - runtime version: " + toolchain.getJavaRuntimeVersion());
-                    }
-                }
+                diags.add("   - vendor: " + toolchain.getVendor());
+                diags.add("   - runtime version: " + toolchain.getJavaRuntimeVersion());
             }
             if (executablePath != null) {
                 try {

--- a/native-gradle-plugin/src/testFixtures/groovy/org/graalvm/buildtools/gradle/fixtures/AbstractFunctionalTest.groovy
+++ b/native-gradle-plugin/src/testFixtures/groovy/org/graalvm/buildtools/gradle/fixtures/AbstractFunctionalTest.groovy
@@ -58,6 +58,7 @@ abstract class AbstractFunctionalTest extends Specification {
 
     String gradleVersion = testGradleVersion()
     boolean debug
+    boolean hasConfigurationCache = Boolean.getBoolean("config.cache")
 
     boolean IS_WINDOWS = System.getProperty("os.name", "unknown").contains("Windows");
     boolean IS_LINUX = System.getProperty("os.name", "unknown").contains("Linux");
@@ -128,7 +129,13 @@ abstract class AbstractFunctionalTest extends Specification {
     void run(String... args) {
         try {
             result = newRunner(args)
-                    .build()
+                    .run()
+            if (hasConfigurationCache) {
+                // run a 2d time to check that not only we can store in
+                // the configuration cache, but that we can also load from it
+                result = newRunner([*args, "--rerun-tasks"] as String[])
+                        .run()
+            }
         } finally {
             recordOutputs()
         }
@@ -182,7 +189,7 @@ abstract class AbstractFunctionalTest extends Specification {
         List<String> autoArgs = [
                 "-S",
         ]
-        if (Boolean.getBoolean("config.cache")) {
+        if (hasConfigurationCache) {
             autoArgs << '--configuration-cache'
         }
         autoArgs << "-I"
@@ -215,7 +222,7 @@ abstract class AbstractFunctionalTest extends Specification {
             allprojects {
                 repositories {
                     maven {
-                        url = "\${providers.systemProperty('common.repo.url').forUseAtConfigurationTime().get()}"
+                        url = "\${providers.systemProperty('common.repo.url').get()}"
                     }
                     mavenCentral()
                 }
@@ -248,7 +255,7 @@ abstract class AbstractFunctionalTest extends Specification {
 
         void contains(String... tasks) {
             tasks.each { task ->
-                assert result.task(task) != null: "Expected to find task $task in the graph but it was missing"
+                assert result.task(task) != null: "Expected to find task $task in the graph but it was missing. Found tasks: ${result.tasks.collect { it.path }}"
             }
         }
 

--- a/samples/java-application-with-custom-tests/build.gradle
+++ b/samples/java-application-with-custom-tests/build.gradle
@@ -53,7 +53,6 @@ application {
 }
 
 def junitVersion = providers.gradleProperty('junit.jupiter.version')
-    .forUseAtConfigurationTime()
     .get()
 
 

--- a/samples/java-application-with-reflection/build.gradle
+++ b/samples/java-application-with-reflection/build.gradle
@@ -53,7 +53,6 @@ application {
 }
 
 def junitVersion = providers.gradleProperty('junit.jupiter.version')
-        .forUseAtConfigurationTime()
         .get()
 
 dependencies {

--- a/samples/java-application-with-resources/build.gradle
+++ b/samples/java-application-with-resources/build.gradle
@@ -53,7 +53,6 @@ application {
 }
 
 def junitVersion = providers.gradleProperty('junit.jupiter.version')
-        .forUseAtConfigurationTime()
         .get()
 
 dependencies {

--- a/samples/java-application-with-tests/build.gradle
+++ b/samples/java-application-with-tests/build.gradle
@@ -53,7 +53,6 @@ application {
 }
 
 def junitVersion = providers.gradleProperty('junit.jupiter.version')
-    .forUseAtConfigurationTime()
     .get()
 
 dependencies {

--- a/samples/java-library/build.gradle
+++ b/samples/java-library/build.gradle
@@ -49,7 +49,6 @@ repositories {
 }
 
 def junitVersion = providers.gradleProperty('junit.jupiter.version')
-    .forUseAtConfigurationTime()
     .get()
 
 dependencies {

--- a/samples/kotlin-application-with-tests/build.gradle
+++ b/samples/kotlin-application-with-tests/build.gradle
@@ -40,7 +40,7 @@
  */
 
 plugins {
-    id 'org.jetbrains.kotlin.jvm' version '1.5.10'
+    id 'org.jetbrains.kotlin.jvm' version '1.7.22'
     id 'application'
     id 'org.graalvm.buildtools.native'
 }

--- a/samples/multi-project-with-tests/core/build.gradle
+++ b/samples/multi-project-with-tests/core/build.gradle
@@ -53,7 +53,6 @@ application {
 }
 
 def junitVersion = providers.gradleProperty('junit.jupiter.version')
-        .forUseAtConfigurationTime()
         .get()
 
 dependencies {

--- a/samples/native-config-integration/build.gradle
+++ b/samples/native-config-integration/build.gradle
@@ -53,7 +53,6 @@ application {
 }
 
 def junitVersion = providers.gradleProperty('junit.jupiter.version')
-        .forUseAtConfigurationTime()
         .get()
 
 dependencies {
@@ -86,19 +85,19 @@ graalvmNative {
 tasks.register("configZip", Zip) {
     destinationDirectory = project.layout.projectDirectory
     from(file("config-directory"))
-    archiveName = 'config-directory.zip'
+    archiveFileName = 'config-directory.zip'
 }
 
 tasks.register("configTarGz", Tar) {
     destinationDirectory = project.layout.projectDirectory
     compression = Compression.GZIP
     from(file("config-directory"))
-    archiveName = 'config-directory.tar.gz'
+    archiveFileName = 'config-directory.tar.gz'
 }
 
 tasks.register("configTarBz2", Tar) {
     destinationDirectory = project.layout.projectDirectory
     compression = Compression.BZIP2
     from(file("config-directory"))
-    archiveName = 'config-directory.tar.bz2'
+    archiveFileName = 'config-directory.tar.bz2'
 }

